metrics/serve.go

```diff
@@ -198,8 +198,35 @@ func handleGradualChangeMode(metricCount, metricLength, metricCycle, seriesCycle
 	}
 }
 
+func handleSpikeMode(metricCount, metricLength, metricCycle, seriesCycle int, labelKeys, labelValues []string, currentSeriesCount *int, spikeMultiplier float64, changeSeriesChan <-chan time.Time, updateNotify chan struct{}) {
+	inSpike := false
+	initialSeriesCount := *currentSeriesCount
+	for tick := range changeSeriesChan {
+		metricsMux.Lock()
+		unregisterMetrics()
+		registerMetrics(metricCount, metricLength, metricCycle, labelKeys)
+		cycleValues(labelKeys, labelValues, *currentSeriesCount, seriesCycle)
+		metricsMux.Unlock()
+
+		if inSpike {
+			inSpike = false
+			*currentSeriesCount = initialSeriesCount
+		} else {
+			inSpike = true
+			*currentSeriesCount = int(float64(initialSeriesCount) * spikeMultiplier)
+		}
+
+		fmt.Printf("%v: Adjusting series count. New count: %d\n", tick, *currentSeriesCount)
+
+		select {
+		case updateNotify <- struct{}{}:
+		default:
```

Comment:
not sure what's happening here, and looking at the other `handle` functions we're doing the same, what do we need this select and the update notify channel for?

Reply:
the updateNotify channel are used here to:
https://github.com/prometheus-community/avalanche/blob/4b732e1a44668d1ed3fa6b503ee2c750d22990ff/metrics/write.go#L140-L149
Each time a notification is received, the metrics are updated and logged.

Reply:
You're right! We learned something. It looks like `updateNotify` is only used for the write mode, where avalanche is remote writing somewhere. So this isn't needed for what we're trying to do with prombench, but it is required in general :+1: